1-Draft/RFCNNNN-RFC-Native-Command-Exit-Errors.md

```diff
@@ -0,0 +1,236 @@
+dynamic scope of a tested context rather than the lexical scope of the
+context. This limits the usefulness of `set -e`, but is needed for
+[historic compatibility](http://austingroupbugs.net/view.php?id=52)
+reasons.
+
+### Treat errors as exceptions only on untested commands
+
+This approach varies the basic specification so that errors in all
+commands can be treated exceptions where the command result is not
+tested.
+
+As well as the new boolean preference to enable reporting errors for
+non-zero native command exit codes, a new preference value option would
+be added for `$ErrorActionPreference`.
+
+A `"StopOnDiscarded"` value for `$ErrorActionPreference` would have the
+effect of ignoring non-terminating command errors in a context where
+the result is tested, and throwing an exception in other contexts.
+
+This should improve on the basic specification by allowing the idiom to
+be combined with normal control flow statement usage and applied to
+cmdlets also.
+
+Treating non-zero exit codes on native commands as errors only in
+untested contexts as with earlier listed approaches gives inconsitent
+error handling between native commands and cmdlets. The parameter
+`-ErrorAction SilentlyContinue` would need to be given to a cmdlet to
+prevent an exception being thrown where `$ErrorActionPreference` is
+`"stop"` and a non-terminating error should not cause an exception
+because the output is being tested.
+
+### Allow output of pipeline to be treated as tested
+
+This approach varies the above approaches to "untested" commands so
+that the output of a pipeline can be treated as tested. 
+
+A `$PSIncludeNativeCommandInErrorActionPreference` variable with three
+possible values would be used instead of a boolean variable. 
+Possible values would be:
+`"none"`:(default) do not report an error on a non-zero exit code
+`"command"`: report an error on a non-zero exit code after a
+native command in an untested pipeline.
+`"pipeline"`: report an error on a non-zero exit code after an
+native command which is the last command in an untested pipeline.
+
+This should add some flexibility for handling errors in multiple
+command pipelines to deal with different expectations.
+
+In Powershell, the status of a pipeline is false if any command in the
+pipeline fails. Thus treating a command that outputs to a pipeline as
+"untested" is arguably be closer to this existing Powershell idiom.
+Korn compatible shells like bash provide similar behaviour with 
+`set -o pipefail`. This makes the exit code of the pipeline zero where
+all exit codes were zero, or otherwise the exit code of the last
+command with a non-zero exit code.
+
+However this might not be the expected behaviour in all circumstances,
+as `set -o pipefail` is not enabled on bourne type shells, and commands
+that output into a pipe might sometimes be considered to have their 
+success or failure "tested" for error handling purposes.
+
+### Use dynamic scope/Set-StrictMode
```

Comment:
While using Version for OS checks does seem wrong, I would love a single command to cover all these:
```
Set-StrictMode -Version 2
$PSIncludeNativeCommandInErrorActionPreference = $true
$ErrorActionPreference = 'Stop'
```

Currently I already put 2 of them on top of every script I write.

While having to put 3 of them is still better than the boilerplate checks, I would love something shorter, e.g.
```
Set-StrictMode -Version 2 -StopOnAllErrors
```

Reply:
It wouldn't be possible to do exactly that, but perhaps something like:
`Set-StrictMode -Version 3 -StopOnAllErrors`
Implementing error handling for native errors would be treated as a strictness level.

Reply:
Do you have a view on wether strictmode versions should be seen strictly (excuse the pun) as governing "language" strictness?
Versions exist for backward compatibility, where a result from "incorrect" code might be better than no result, pending corrections to the code. By that measure, enabling error detection on code that previously gave a "successful" but unreliable results seems quite well aligned with strictness versions?
Or would would your case be better described with something like:
`Set-StrictMode -Version 2 -NativeErrrorVersion 1 -StopOnAllErrors`

Reply:
I'm not sure -- it does feel language-related to me, but I think it's clear either way. Might be worth asking people who are currently using `Set-StrictMode`, but not using `$ErrorActionPreference`, though I don't know of any.

My goal is to have something very short, e.g. `Set-StrictMode -Version 2 -StopOnAllErrors` (with `All` meaning native and non-native) or `Set-StrictMode -Version 3`. People who need some special case can use `$PSIncludeNativeCommandInErrorActionPreference` directly, but I think a common case of maximum strictness should be short and easy.

Reply:
Thanks, I don't know what the numbers are either way. I would always set strictmode as it's about error detection and 'strict' correctness, but control $ErrorActionPreference separately so I can adjust error handling for testing etc., but I think I see why you something like "throw on all errors" might be useful.